### PR TITLE
Fixes export destructured variables reference

### LINKED ISSUE #31922: tsserver does'nt output references that array desturctured variables
**TypeScript Version:** 3.6.0-dev.20190618

**Search Terms:**

`tsserver` `references` `peek references` `array destructuring assignment`

**Code**

index.ts

```ts
import { a, b, c } from './mod';
console.log(a, b, c);
```

mod.ts

```ts
export const [a, b] = [1, 2];
export const c = 3;
```

- run following script that is to get variable `b`'s references

```console

npx tsserver <<EOF | tail -n 1
{"command":"open","arguments":{"file":"/path/to/mod.ts"}}
{"command":"references","arguments":{"file":"/path/to/mod.ts","line":1,"offset":19}}
EOF
```

**Expected behavior:**

tsserver returned all variable `b`'s references( `refs` field should contain `index.ts`'s result), like following
script(to get variable `c`'s references) result

```json
{"seq":0,"type":"response","command":"references","success":true,"body":{"refs":[{"file":"/path/to/ts-array-destructured-refs/src/mod.ts","start":{"line":2,"offset":14},"end":{"line":2,"offset":15},"lineText":"export const c = 3;","isWriteAccess":true,"isDefinition":true},{"file":"/path/to/ts-array-destructured-refs/src/index.ts","start":{"line":1,"offset":16},"end":{"line":1,"offset":17},"lineText":"import { a, b, c } from './mod';","isWriteAccess":true,"isDefinition":true},{"file":"/path/to/ts-array-destructured-refs/src/index.ts","start":{"line":2,"offset":19},"end":{"line":2,"offset":20},"lineText":"console.log(a, b, c);","isWriteAccess":false,"isDefinition":false}],"symbolName":"c","symbolStartOffset":14,"symbolDisplayString":"const c: 3"}}
```

**Actual behavior:**

tsserver returned `refs` field that contain `index.ts`'s variables.

result output

```json
{"seq":0,"type":"response","command":"references","success":true,"body":{"refs":[{"file":"/path/to/ts-array-destructured-refs/src/mod.ts","start":{"line":1,"offset":18},"end":{"line":1,"offset":19},"lineText":"export const [a, b] = [1, 2];","isWriteAccess":true,"isDefinition":true}],"symbolName":"b","symbolStartOffset":18,"symbolDisplayString":"const b: number"}}
```

**Playground Link:**

https://github.com/sisisin-sandbox/ts-array-destructured-refs

### PR DESCRIPTION
**Pull request by sisisin**
Fixes #31922

### CODE REVIEW

File: src/services/importTracker.ts [hunk 1]
```diff
@@ -582,6 +582,12 @@ namespace ts.FindAllReferences {
             return p.name !== node ? undefined :
                 p.parent.kind === SyntaxKind.CatchClause ? undefined : p.parent.parent.kind === SyntaxKind.VariableStatement ? p.parent.parent : undefined;
         }
+        else if (parent.parent && parent.parent.parent && parent.parent.parent.kind === SyntaxKind.VariableDeclaration) {
```
Comment (ajafff) [inline src/services/importTracker.ts]: does this work for binding elements of arbitrary nesting?

```ts
export const {foo: [bar, {baz: bas}]} = someObj;
```

You probably want to use `walkUpBindingElementsAndPatterns` to get the VariableDeclaration of the BindingElement.

File: src/testRunner/unittests/tsserver/getExportReferences.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    describe("unittests:: tsserver:: getExportReferences", () => {
+        const mainTs: File = {
+            path: "/main.ts",
+            content: 'import { value, valueA, valueB, valueC, valueD } from "./mod";',
+        };
+        const modTs: File = {
+            path: "/mod.ts",
+            content: `export const value = 0;
+export const [valueA, valueB] = [0, 1];
+export const { valueC, valueD } = { valueC: 0, valueD: 1 };`,
```
Comment (ajafff) [inline src/testRunner/unittests/tsserver/getExportReferences.ts]: can you add a test that renames the destructured property `export const {foo: bar} = {foo: 2}` to make sure this works as expected

**Comment by sisisin** [comment]
@ajafff Thank you for your review, I fixed them

File: src/testRunner/unittests/tsserver/getExportReferences.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+            override: Partial<MakeReferenceItem> = {},
+        ): protocol.ReferencesResponseItem =>
+            makeReferenceItem({
+                file: modTs,
+                isDefinition: true,
+                ...texts,
+                ...override,
+            });
+
+        it("should get const variable decralation references", () => {
```
Comment (rbuckton) [inline src/testRunner/unittests/tsserver/getExportReferences.ts]:
```suggestion
        it("should get const variable declaration references", () => {
```

File: src/testRunner/unittests/tsserver/getExportReferences.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                ],
+                symbolDisplayString: "const value: 0",
+                symbolName: "value",
+                symbolStartOffset: protocolLocationFromSubstring(modTs.content, "value").offset,
+            };
+
+            assert.deepEqual(response, expectResponse);
+        });
+
+        it("should get array destructuring decralation references", () => {
```
Comment (rbuckton) [inline src/testRunner/unittests/tsserver/getExportReferences.ts]:
```suggestion
        it("should get array destructuring declaration references", () => {
```

File: src/testRunner/unittests/tsserver/getExportReferences.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                ],
+                symbolDisplayString: "const valueA: number",
+                symbolName: "valueA",
+                symbolStartOffset: protocolLocationFromSubstring(modTs.content, "valueA").offset,
+            };
+
+            assert.deepEqual(response, expectResponse);
+        });
+
+        it("should get object destructuring decralation references", () => {
```
Comment (rbuckton) [inline src/testRunner/unittests/tsserver/getExportReferences.ts]:
```suggestion
        it("should get object destructuring declaration references", () => {
```

File: src/testRunner/unittests/tsserver/getExportReferences.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+                ],
+                symbolDisplayString: "const valueC: number",
+                symbolName: "valueC",
+                symbolStartOffset: protocolLocationFromSubstring(modTs.content, "valueC").offset,
+            };
+
+            assert.deepEqual(response, expectResponse);
+        });
+
+        it("should get object decralation references that renames destructured property", () => {
```
Comment (rbuckton) [inline src/testRunner/unittests/tsserver/getExportReferences.ts]:
```suggestion
        it("should get object declaration references that renames destructured property", () => {
```

File: src/testRunner/unittests/tsserver/getExportReferences.ts [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+                ],
+                symbolDisplayString: "const renamedD: number",
+                symbolName: "renamedD",
+                symbolStartOffset: protocolLocationFromSubstring(modTs.content, "renamedD").offset,
+            };
+
+            assert.deepEqual(response, expectResponse);
+        });
+
+        it("should get nested object decralation references", () => {
```
Comment (rbuckton) [inline src/testRunner/unittests/tsserver/getExportReferences.ts]:
```suggestion
        it("should get nested object declaration references", () => {
```

**Comment by sisisin** [comment]
@rbuckton Thank you for your review, I fixed each commits to use rebase and edit.

**Comment by rbuckton** [comment]
Thanks for the contribution!

**Pull request merged**